pkg/lib/cockpit/channel.ts

```diff
@@ -0,0 +1,240 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2024 Red Hat, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+import { JsonObject } from './_internal/common';
+import { Transport, ensure_transport, transport_globals } from './_internal/transport';
+import { EventEmitter } from './event';
+
+type ChannelPayload = string | Uint8Array;
+
+type ChannelBinaryOption<P extends ChannelPayload> =
+    P extends Uint8Array ?
+        { binary: true }
+    : P extends string ?
+        { binary?: false }
+    :
+        { binary: boolean }
+    ;
+
+type ChannelOptions<P extends ChannelPayload> = JsonObject & ChannelBinaryOption<P> & {
+    command?: never;
+    channel?: never;
+    host?: string;
+    payload: string;
+    superuser?: "try" | "require";
+};
+
+interface ChannelControlMessage extends JsonObject {
+    command: string;
+}
+
+interface ChannelEvents<P extends ChannelPayload> {
+    control(options: ChannelControlMessage): void;
+    done(options: ChannelControlMessage): void;
+    ready(options: ChannelControlMessage): void;
+    close(options: ChannelControlMessage): void;
+    data(data: P): void;
+}
+
+export class Channel<P extends ChannelPayload = ChannelPayload> extends EventEmitter<ChannelEvents<P>> {
+    _payload_type: P | null = null; // required to forbid mixing types
+
+    id: string | null = null; // can be unassigned during transport startup
+    readonly options: ChannelOptions<P>;
+    readonly binary: ChannelBinaryOption<P>['binary'];
+
+    #transport: Transport | null = null;
+    #received: Partial<Record<"close" | "ready" | "done", ChannelControlMessage>> = {};
+    #queue: ([true, ChannelControlMessage] | [false, P])[] = [];
+
+    on_control(control: ChannelControlMessage) {
+        const command = control.command;
+
+        if (command === 'ready' || command === 'close' || command === 'done') {
+            if (command in this.#received) {
+                return;
+            }
+
+            this.#received[command] = control;
+            this.emit(command, control);
+        } else {
+            this.emit('control', control);
```

Comment:
I checked a lot of the umpteen uncovered code complaints. Many are really unreachable "should not happen" code, and things like this are because we just ported over one channel type. This should get covered if we e.g. port the file channel to this (anything that supports a binary channel).